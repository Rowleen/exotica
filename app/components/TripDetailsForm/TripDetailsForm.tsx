import { FC, useEffect, useState, useMemo } from 'react'
import Image from 'next/image'
import useCreateTrip from '../../core/domain/useCases/useCreateTrip'
import useUpdateTrip from '../../core/domain/useCases/useUpdateTrip'
import Input from '../Input/Input'
import Textarea from '../Textarea/Textarea'
import Button from '../Button/Button'
import ItineraryDay from '../ItineraryDay/ItineraryDay'

import type { Trip } from '../../core/domain/entities/Trip'
import type { onChangeEvent } from '../../shared/types/onChangeEvent'

import styles from './tripDetailsForm.module.sass'

interface TripDetailsProps {
  trip?: Trip
  action?: 'create' | 'edit'
}

const TripDetailsForm: FC<TripDetailsProps> = ({ trip, action }) => {
  const [formData, setFormData] = useState<Trip>({
    id: 0,
    title: '',
    description: '',
    photo_url: '',
    status: 'todo',
    itinerary: [
      {
        day: 1,
        location: '',
        description: ''
      }
    ]
  })

  const { mutate: createTrip } = useCreateTrip()
  const { mutate: updateTrip } = useUpdateTrip()

  useEffect(() => {
    if (trip && action === 'edit') {
      setFormData(trip)
    }
  }, [action, trip])

  const handleInputChange = (event: onChangeEvent, index?: number) => {
    const { name, value } = event.target

    // @ts-expect-error: it's evaluating no falsy result from 0
    if (typeof index !== false && index !== undefined) {
      return setFormData({
        ...formData,
        itinerary: formData.itinerary.map((day, i) =>
          i === index ? { ...day, [name]: value } : day
        )
      })
    }

    setFormData({
      ...formData,
      [name]: value
    })
  }

  const handleAddItineraryDay = () => {
    const day = {
      day: 1,
      location: '',
      description: ''
    }

    setFormData({
      ...formData,
      itinerary: [...formData.itinerary, day]
    })
  }

  const handleSubmit = (event: React.MouseEvent) => {
    event.preventDefault()

    if (action === 'create') {
      createTrip(formData)
    }

    if (action === 'edit') {
      updateTrip(formData)
    }
  }

  const startTitle = useMemo(() => {
    if (action === 'create') {
      return 'Create'
    }

    if (action === 'edit') {
      return 'Edit'
    }
  }, [action])

  return (
    <section className={styles.wrapper}>
      <h2 className={styles.title}>{`${startTitle} a trip`}</h2>

      <form>
        <div className={styles.inputGroup}>
          <label className={styles.label}>Name*</label>
          <Input
            name='title'
            type='text'
            placeholder='Country'
            onChange={event => handleInputChange(event)}
            value={formData.title}
            required
          />
        </div>

        <div className={styles.inputGroup}>
          <label className={styles.label}>
            Description (max. 240 characters)
          </label>
          <Textarea
            name='description'
            rows={5}
            placeholder='From Rome to Venice'
            onChange={event => handleInputChange(event)}
            value={formData.description}
            required
            maxLength={240}
          />
        </div>

        <div className={styles.inputGroup}>
          <label className={styles.label}>Image</label>
          <Input
            name='photo_url'
            type='url'
            placeholder='Image URL'
            onChange={event => handleInputChange(event)}
            value={formData.photo_url}
            pattern='[Hh][Tt][Tt][Pp][Ss]?:\/\/(?:(?:[a-zA-Z\u00a1-\uffff0-9]+-?)*[a-zA-Z\u00a1-\uffff0-9]+)(?:\.(?:[a-zA-Z\u00a1-\uffff0-9]+-?)*[a-zA-Z\u00a1-\uffff0-9]+)*(?:\.(?:[a-zA-Z\u00a1-\uffff]{2,}))(?::\d{2,5})?(?:\/[^\s]*)?'
          />
        </div>

        <div className={styles.itineraryHeader}>
          <label className={styles.label}>Day by day itinerary</label>

          <Image
            className={styles.iconAdd}
            src='/images/plus-circle.svg'
            alt='Add icon'
            width={0}
            height={0}
            role='button'
            onClick={handleAddItineraryDay}
          />
        </div>

        {formData.itinerary.map((day, index) => (
          <ItineraryDay
            key={day.location + index}
            value={formData.itinerary[index]}
            onChange={event => handleInputChange(event, index)}
          />
        ))}

        <Button
          text='Save'
          type='submit'
          shape='button'
          size='medium'
          color='primary'
          onClick={event => handleSubmit(event)}
        />
      </form>
    </section>
  )
}

export default TripDetailsForm
