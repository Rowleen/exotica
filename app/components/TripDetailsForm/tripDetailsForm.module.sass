@import '../../sass/variables'

.wrapper
  width: 100%
  display: flex
  flex-direction: column
  padding: $spacing-4xl $spacing-3xl

  .title
    color: $font-color-primary
    font-family: 'Roboto Bold'
    margin-bottom: $spacing-3xl

  .inputGroup
    margin-bottom: $spacing-3xl

    .label
      display: block
      color: $font-color-primary
      font-size: 0.87rem
      font-family: 'Roboto Regular'
      margin-bottom: $spacing-xs

  .itineraryHeader
    display: flex
    margin-bottom: $spacing-xs

    .label
      display: block
      flex: 1
      color: $font-color-primary
      font-size: 0.87rem
      font-family: 'Roboto Regular'

    .iconAdd
      cursor: pointer
      width: 21px
      height: 21px