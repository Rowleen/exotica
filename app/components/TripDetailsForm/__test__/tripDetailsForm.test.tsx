import React from 'react'
import '@testing-library/jest-dom'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import TripDetailsForm from '../TripDetailsForm'

const queryClient = new QueryClient()

import tripMock from '../../../__mocks__/trip.mock.json'
import tripWithNoItineraryMock from '../../../__mocks__/tripWithNoItinerary.mock.json'

describe('Home', () => {
  const renderComponent = (
    <QueryClientProvider client={queryClient}>
      <TripDetailsForm trip={tripMock} action='edit' />
    </QueryClientProvider>
  )

  it('renders trip details form modal unchanged', () => {
    const { container } = render(renderComponent)
    expect(container).toMatchSnapshot()
  })

  it('render country name', () => {
    const { getByDisplayValue } = render(renderComponent)

    const titleField = getByDisplayValue(tripMock.title)

    expect(titleField).toBeInTheDocument()
  })

  it('render trip description', () => {
    const { getByDisplayValue } = render(renderComponent)

    const descriptionField = getByDisplayValue(tripMock.description)

    expect(descriptionField).toBeInTheDocument()
  })

  it('render trip description', () => {
    const { getByDisplayValue } = render(renderComponent)

    const descriptionField = getByDisplayValue(tripMock.description)

    expect(descriptionField).toBeInTheDocument()
  })

  it('render trip photo url', () => {
    const { getByDisplayValue } = render(renderComponent)

    const photoUrlField = getByDisplayValue(tripMock.photo_url)

    expect(photoUrlField).toBeInTheDocument()
  })

  it('render two days of itinerary', () => {
    const { getAllByLabelText } = render(renderComponent)

    const itinerary = getAllByLabelText('location')

    expect(itinerary).toHaveLength(tripMock.itinerary.length)
  })

  test("no render itineray when it won't be created by user", () => {
    const { queryByLabelText } = render(
      <QueryClientProvider client={queryClient}>
        <TripDetailsForm trip={tripWithNoItineraryMock} action='edit' />
      </QueryClientProvider>
    )

    const itinerary = queryByLabelText('location')

    expect(itinerary).toBeNull
  })
})
