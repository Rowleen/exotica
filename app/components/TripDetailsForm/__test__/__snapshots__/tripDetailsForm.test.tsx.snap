// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders trip details form modal unchanged 1`] = `
<div>
  <section
    class="wrapper"
  >
    <h2
      class="title"
    >
      Edit a trip
    </h2>
    <form>
      <div
        class="inputGroup"
      >
        <label
          class="label"
        >
          Name*
        </label>
        <input
          aria-label="title"
          class="input"
          name="title"
          placeholder="Country"
          required=""
          type="text"
          value="Portugal"
        />
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="label"
        >
          Description (max. 240 characters)*
        </label>
        <textarea
          aria-label="description"
          class="textarea"
          maxlength="240"
          name="description"
          placeholder="From Rome to Venice"
          required=""
          rows="5"
        >
          Embark on a journey through Portugal, where the charming streets of Lisbon captivate you, the golden beaches of the Algarve await, and Portuguese cuisine delights with authentic flavors. Explore castles in Sintra and create unforgettable memories in this destination full of history and beauty. Portugal invites you to experience something truly unique!
        </textarea>
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="label"
        >
          Image
        </label>
        <input
          aria-label="photo_url"
          class="input"
          name="photo_url"
          pattern="[Hh][Tt][Tt][Pp][Ss]?:\\/\\/(?:(?:[a-zA-Z\\u00a1-\\uffff0-9]+-?)*[a-zA-Z\\u00a1-\\uffff0-9]+)(?:\\.(?:[a-zA-Z\\u00a1-\\uffff0-9]+-?)*[a-zA-Z\\u00a1-\\uffff0-9]+)*(?:\\.(?:[a-zA-Z\\u00a1-\\uffff]{2,}))(?::\\d{2,5})?(?:\\/[^\\s]*)?"
          placeholder="Image URL"
          type="url"
          value="https://a.cdn-hotels.com/gdcs/production82/d1923/447a348f-f875-4885-b00a-e9a90603fef5.jpg"
        />
      </div>
      <div
        class="itineraryHeader"
      >
        <label
          class="label"
        >
          Day by day itinerary
        </label>
        <img
          alt="Add icon"
          class="iconAdd"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          role="button"
          src="/images/plus-circle.svg"
          style="color: transparent;"
          width="0"
        />
      </div>
      <section
        class="wrapper"
      >
        <div
          class="day"
        >
          <div
            class="leftColumn"
          >
            <select
              aria-label="day"
              class="select"
              name="day"
            >
              <option
                disabled=""
              >
                Day
              </option>
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
              <option
                value="11"
              >
                11
              </option>
              <option
                value="12"
              >
                12
              </option>
              <option
                value="13"
              >
                13
              </option>
              <option
                value="14"
              >
                14
              </option>
              <option
                value="15"
              >
                15
              </option>
            </select>
          </div>
          <div
            class="rightColumn"
          >
            <input
              aria-label="location"
              class="input"
              name="location"
              placeholder="Location"
              type="text"
              value="Lisbon"
            />
            <textarea
              aria-label="itineraryDescription"
              class="textarea"
              name="description"
              placeholder="Description"
              rows="5"
            >
              Explore the Alfama neighborhood and visit São Jorge Castle.
            </textarea>
          </div>
        </div>
      </section>
      <section
        class="wrapper"
      >
        <div
          class="day"
        >
          <div
            class="leftColumn"
          >
            <select
              aria-label="day"
              class="select"
              name="day"
            >
              <option
                disabled=""
              >
                Day
              </option>
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
              <option
                value="11"
              >
                11
              </option>
              <option
                value="12"
              >
                12
              </option>
              <option
                value="13"
              >
                13
              </option>
              <option
                value="14"
              >
                14
              </option>
              <option
                value="15"
              >
                15
              </option>
            </select>
          </div>
          <div
            class="rightColumn"
          >
            <input
              aria-label="location"
              class="input"
              name="location"
              placeholder="Location"
              type="text"
              value="Lisbon"
            />
            <textarea
              aria-label="itineraryDescription"
              class="textarea"
              name="description"
              placeholder="Description"
              rows="5"
            >
              Visit the Jerónimos Monastery and the Monument to the Discoveries.
            </textarea>
          </div>
        </div>
      </section>
      <button
        class="button primary medium disabled"
        disabled=""
        type="submit"
      >
        Save
      </button>
    </form>
  </section>
</div>
`;
